main/core/src/eval/Evaluator.scala

```diff
@@ -362,15 +382,302 @@ case class Evaluator(home: os.Path,
       }
     }
 
+    logEnd.foreach(_())
+
     multiLogger.close()
 
     (newResults, newEvaluated)
   }
 
   def resolveLogger(logPath: Option[os.Path], logger: Logger): Logger = logPath match{
     case None => logger
-    case Some(path) => MultiLogger(logger.colored, log, FileLogger(logger.colored, path, debugEnabled = true))
+    case Some(path) => MultiLogger(logger.colored, log, new FileLogger(logger.colored, path, debugEnabled = true))
   }
+
+  type Terminal = Either[Task[_], Labelled[Any]]
+  type TerminalGroup = (Terminal, Agg[Task[_]])
+
+  def evaluateParallel(
+    goals: Agg[Task[_]],
+    threadCount: Int,
+    reporter: Int => Option[BuildProblemReporter] = (int: Int) => Option.empty[BuildProblemReporter],
+    testReporter: TestReporter = DummyTestReporter,
+    logger: Logger = log
+  ): Evaluator.Results = {
+    os.makeDir.all(outPath)
+
+    val startTime = System.currentTimeMillis()
+
+    // we need to collect all relevant logging info while developing
+    val evalLog = MultiLogger(true, this.log,
+      new FileLogger(false, outPath / "evaluator.log", true, append = true) {
+        override def debug(s: String) = super.debug(s"${System.currentTimeMillis() - startTime} [${Thread.currentThread().getName()}] ${s}")
+      })
+
+    evalLog.info(s"Using experimental parallel evaluator with ${threadCount} threads")
+    evalLog.debug(s"Start time: ${new java.util.Date()}")
+
+    val (sortedGroups, transitive) = Evaluator.plan(rootModule, goals)
+
+    // Mutable collector for all evaluated tasks
+    val evaluated = new Agg.Mutable[Task[_]]
+
+    // Mutable collector for all task results
+    @volatile var results = Map.empty[Task[_], Result[(Any, Int)]]
```

Comment:
Does this really need to be volatile? I notice the other mutable data structures all seem to be vanilla, e.g. `timings` below and `evaluated` above. I also don't see `results` being used anywhere

Reply:
`results` is also read inside a future to have access to all results.

Reply:
I changed it in favour to use a mutable ConcurrentHashMap.